src/items/enumerations.md

```diff
@@ -56,22 +56,64 @@ a = Animal::Cat { name: "Spotty".to_string(), weight: 2.7 };
 ```
 
 In this example, `Cat` is a _struct-like enum variant_, whereas `Dog` is simply
-called an enum variant. Each enum instance has a _discriminant_ which is an
-integer associated to it that is used to determine which variant it holds. An
-opaque reference to this discriminant can be obtained with the
-[`mem::discriminant`] function.
+called an enum variant.
 
-## Custom Discriminant Values for Fieldless Enumerations
+An enum where no constructors contain fields are called a
+*<a name="field-less-enum">field-less enum</a>*.
 
-If there is no data attached to *any* of the variants of an enumeration,
-then the discriminant can be directly chosen and accessed.
+## Discriminants
 
-These enumerations can be cast to integer types with the `as` operator by a
-[numeric cast]. The enumeration can optionally specify which integer each
-discriminant gets by following the variant name with `=` followed by a [constant
-expression]. If the first variant in the declaration is unspecified, then it is
-set to zero. For every other unspecified discriminant, it is set to one higher
-than the previous variant in the declaration.
+Each enum instance has a _discriminant_: an integer logically associated to it
+that is used to determine which variant it holds.
+
+Under the [default representation], the discriminant is interpreted as
+an `isize` value. However, the compiler is allowed to use a smaller type (or
+another means of distinguishing variants) in its actual memory layout.
+
+### Assigning Discriminant Values
+
+#### Explicit Discriminants
+
+In two circumstances, the discriminant of a variant may be explicitly set by
+following the variant name with `=` and a [constant expression]:
+
+<ol>
+<li>
+
+if the enumeration is "C-like" (i.e., it has no tuple or struct variants); e.g.:
```

Comment:
```suggestion
if the enumeration is fieldless (i.e., it has no tuple or struct variants); e.g.:
```

Reply:
This is wrong because the compiler would tell me to add a repr for the enum.

Reply:
I don't understand what this has to do with the representation.  We have been moving the reference to use the term "fieldless enum" instead of "c-like"; that is the terminology we would prefer.

Reply:
True, but only c-like enums are allowed to have a discriminant without a `#[repr()]` attribute.

Reply:
Sorry, I'm still confused.  The term "c-like enum" has historically been used for Rust enums that only has unit variants (that is, variants without fields).  We have changed the terminology in the reference to use the phrase "fieldless enum" instead.  We want to stop using the phrase "c-like" altogether.  Indeed, this particular section is showing a fieldless enum.  Why do you still want to use the term "c-like" here?

Reply:
The following "c-like enum" does not need a repr attribute: (this is also "fieldless")
```rust
enum NoRepr {
    Foo = 5,
    Bar = 3,
    Baz = 1,
}
```

While the following "fieldless enum" (but not "c-like") **needs** a `repr` attribute:

```rust
#[repr(u8)] // MUST do this
enum B {
    Foo = 5,
    Bar() = 3, 
    Baz { } = 1,
}
```

Is this intended by the RFC? I think we might allow these kinds of "fieldless" but not "c-like" enums like this without a `repr` attribute.

cc @jswrenn @nox

Reply:
Hm, I can see how the literal interpretation of "fieldless" could be confused with those.  The two examples with brackets define tuple and struct variants, they just have 0 fields.  The unit variant has a distinctly different meaning which has different semantics.  As defined here in the reference, a "fieldless enum" is one with only unit variants (your first example).

Reply:
I don't think that is correct. The reference says 

> An enum where no constructors contain fields are called a
*<a id="field-less-enum">field-less enum</a>*.

Reply:
Well, this PR is the one adding that definition, no?  I guess my review could be more thorough here, but if you want to make that statement more precise, I would reword it to be that a fieldless enum is one that only has unit variants.  I would also drop the hyphen.  AFAIK, the language has always generally treated unit variants different from struct and tuple variants, even if the later two don't have fields.

Reply:
all "fieldless" enums can be cast to their discriminant values. however only c-like enums may specify discriminants without a repr. @jswrenn has already said this on the original PR: https://github.com/rust-lang/reference/pull/639#issuecomment-511515290

Reply:
Opened https://github.com/rust-lang/rust/pull/88203. This should be made consistent and I see no problems for fieldless enums to have an unspecified repr.